### PR TITLE
Add EALSTM 

### PR DESCRIPTION
**Pull request by gabrieltseng**
Adds an entity-aware LSTM, as described here: https://github.com/kratzert/ealstm_regional_modeling

Most of the code is just copied from the EALSTM described there, and fitted to our current RNN model.

Currently, the static data is the latlon data, but this should be updated to include landcover, and eventually topographical data.

### CODE REVIEW

**Comment by gabrieltseng** [comment]
TODO: add mean values

File: src/models/neural_networks/ealstm.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+            nn.init.kaiming_uniform_(dense_layer.linear.weight.data)
+
+        nn.init.kaiming_uniform_(self.final_dense.weight.data)
+        nn.init.constant_(self.final_dense.bias.data, 0)
+
+    def forward(self, x, pred_month=None, latlons=None, current=None):
+
+        # has to be optional to respect the positionality of the neural
+        # nets
+        assert latlons is not None, "Latlons can't be None!"
```
Comment (tommylees112) [inline src/models/neural_networks/ealstm.py]: can we not make latlons a required argument? Or  does the position change make that not possible?

Comment (gabrieltseng) [reply src/models/neural_networks/ealstm.py]: yea - it would screw up the indexing

**Comment by gabrieltseng** [comment]
Added mean values. Specifically:

* **Mean monthly values**: these are (optionally) added to the `current` array
* **Mean yearly values**: these are (optionally) passed as an additional argument to the models. Different models deal with them differently.

File: src/models/data.py [hunk 2]
```diff
@@ -184,15 +194,28 @@ def calculate_normalizing_array(self, data_vars: List[str]):
         for var in data_vars:
             for norm_var in normalizing_dict_keys:
                 if var.endswith(norm_var):
-                    mean.append(self.normalizing_dict[norm_var]['mean'])
-                    std.append(self.normalizing_dict[norm_var]['std'])
+                    var_mean = self.normalizing_dict[norm_var]['mean']
+                    var_std = self.normalizing_dict[norm_var]['std']
+                    if yearly_mean:
+                        var_mean = np.mean(var_mean)
```
Comment (tommylees112) [inline src/models/data.py]: this is definitely across all of the data and not the batches right?

Comment (tommylees112) [reply src/models/data.py]: how does this differ from the `yearly_mean` defined below on L230:

```python
        yearly_mean = x.mean(dim=['time', 'lat', 'lon'])
        yearly_mean_np = yearly_mean.to_array().values

        x = self._add_extra_dims(x, self.surrounding_pixels, self.monthly_means)


```

File: src/models/data.py [hunk 3]
```diff
@@ -234,9 +263,11 @@ def ds_folder_to_np(self,
         y_np = y_np.reshape(y_np.shape[0], y_np.shape[1], y_np.shape[2] * y_np.shape[3])
         y_np = np.moveaxis(y_np, -1, 0).reshape(-1, 1)
 
-        if self.normalizing_array is not None:
+        if (self.normalizing_array is not None) and (self.normalizing_array_ym is not None):
             x_np = (x_np - self.normalizing_array['mean']) / (self.normalizing_array['std'])
             train_latlons = train_latlons / [90, 180]
+            yearly_mean_np = (yearly_mean_np - self.normalizing_array_ym['mean']) / \
```
Comment (tommylees112) [inline src/models/data.py]: so we're not including the yearly mean data we are also normalizing the data by the annual mean?

Does that make sense to duplicate that information? Wouldn't normalisation by the global mean and then including the annual mean have all of that information already incorporated?

Comment (gabrieltseng) [reply src/models/data.py]: `normalizing_array_ym` is the global mean information, fitted to an array that has the same dimensions as the yearly mean.

So normalization only happens using the global mean

File: src/models/data.py [hunk 4]
```diff
@@ -297,24 +332,33 @@ def ds_folder_to_np(self,
...
-    def _add_surrounding(x: xr.Dataset, surrounding_pixels: int) -> xr.Dataset:
-
-        lat_shifts = lon_shifts = range(-surrounding_pixels, surrounding_pixels + 1)
-        for var in x.data_vars:
-            for lat_shift in lat_shifts:
-                for lon_shift in lon_shifts:
-                    if lat_shift == lon_shift == 0: continue
-                    shifted_varname = f'lat_{lat_shift}_lon_{lon_shift}_{var}'
-                    x[shifted_varname] = x[var].shift(lat=lat_shift, lon=lon_shift)
+    def _add_extra_dims(x: xr.Dataset, surrounding_pixels: Optional[int],
```
Comment (tommylees112) [inline src/models/data.py]: This is a great idea. Helps to clean some of the code up

File: src/models/data.py (hunk 2 above)
Comment (gabrieltseng) [reply src/models/data.py]: What happens on line 200 is calculation of the `normalizing array` for the `yearly_mean_array` calculated in line 230

File: tests/models/test_data.py [hunk 5]
```diff
@@ -196,27 +197,47 @@ def __init__(self):
...
-                        org = x_with_more.VHI.isel(time=0, lon=5, lat=5).values
-                        shifted = x_with_more[shifted_var_name].isel(time=0,
-                                                                     lon=5 + lon,
-                                                                     lat=5 + lat).values
-                        assert org == shifted, f"Shifted values don't match!"
+            if surrounding_pixels is not None:
+                for lat in [-1, 0, 1]:
+                    for lon in [-1, 0, 1]:
+                        if lat == lon == 0:
+                            assert f'lat_{lat}_lon_{lon}_{data_var}' not in shifted_mean_vars, \
```
Comment (tommylees112) [inline tests/models/test_data.py]: What does this test for sorry?
* that the surrounding pixels are only included in the historical / `shifted` (shifted in time?) variables

Comment (gabrieltseng) [reply tests/models/test_data.py]: This tests that `lat_1_lon_1_{var}` is equal to the variable, shifted by one latitude and longitude

File: src/models/regression.py [hunk 6]
```diff
@@ -270,3 +215,32 @@ def _calculate_big_mean(self) -> np.ndarray:
             mean * size / total_size for mean, size in zip(means, sizes)
         ]
         return sum(weighted_means)
+
+    def _concatenate_data(self, x: Union[Tuple[Optional[np.ndarray], ...],
```
Comment (tommylees112) [inline src/models/regression.py]: thanks this is a super useful function...

File: tests/models/neural_networks/test_ealstm.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    def test_save(self, tmp_path, monkeypatch):
+
+        features_per_month = 5
+        dense_features = [10]
+        hidden_size = 128
+        rnn_dropout = 0.25
+        include_latlons = True
+        include_pred_month = True
+        include_yearly_means = True
+        yearly_mean_size = 3
```
Comment (tommylees112) [inline tests/models/neural_networks/test_ealstm.py]: Sorry can you remind me why this has a size? Surely the yearly mean is a product of the number of `features_per_month` ?

Comment (gabrieltseng) [reply tests/models/neural_networks/test_ealstm.py]: Yes, but its not trivial to work out what the relationship is (for instance, if there are monthly aggregates or surrounding pixels). So its easier to pass this as an argument to the model directly

**Comment by gabrieltseng** [comment]
Added standard deviations too

File: tests/models/neural_networks/test_ealstm.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        model = EARecurrentNetwork(hidden_size=hidden_size, dense_features=dense_features,
+                                   rnn_dropout=rnn_dropout, data_folder=tmp_path)
+        model.train()
+
+        captured = capsys.readouterr()
+        expected_stdout = 'Epoch 1, train smooth L1: 0.'
+        assert expected_stdout in captured.out
+
+        assert type(model.model) == EALSTM, \
+            f'Model attribute not an RNN!'
```
Comment (tommylees112) [inline tests/models/neural_networks/test_ealstm.py]: tiny change sorry - EALSTM ? or meant to be RNN?

File: src/models/neural_networks/ealstm.py [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+        x_d : torch.Tensor
+            Tensor, containing a batch of sequences of the dynamic features. Shape has to match
+            the format specified with batch_first.
+        x_s : torch.Tensor
+            Tensor, containing a batch of static features.
+        Returns
+        -------
+        h_n : torch.Tensor
+            The hidden states of each time step of each sample in the batch.
+        c_n : torch.Tensor]
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply src/models/neural_networks/ealstm.py]: Ah that looks like a typo - I just copied this from the ealstm repo

File: src/models/neural_networks/ealstm.py [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+            The cell states of each time step of each sample in the batch.
+        """
+        if self.batch_first:
+            x_d = x_d.transpose(0, 1)
+
+        seq_len, batch_size, _ = x_d.size()
+
+        h_0 = x_d.data.new(batch_size, self.hidden_size).zero_()
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply src/models/neural_networks/ealstm.py]: This is making a tensor of all zeros:
https://pytorch.org/docs/stable/tensors.html?highlight=zero_#torch.Tensor.zero_

File: tests/models/test_data.py [hunk 11]
```diff
@@ -197,11 +197,13 @@ def __init__(self):
             assert np.all(x_train_data.current == expected), f"Expected to " \
                 "find the target_time data for the non target variables"
 
-        assert x_train_data.yearly_means.shape[1] == 4
+        assert x_train_data.yearly_aggs.shape[1] == 4 * 2  # means and std
```
Comment (tommylees112) [inline tests/models/test_data.py]: 4 is the number of variables yes?

File: src/models/neural_networks/ealstm.py [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+        # calculate input gate only once because inputs are static
+        bias_s_batch = (self.bias_s.unsqueeze(0).expand(batch_size, *self.bias_s.size()))
+        i = torch.sigmoid(torch.addmm(bias_s_batch, x_s, self.weight_sh))
+
+        # perform forward steps over input sequence
+        for t in range(seq_len):
+            h_0, c_0 = h_x
+
+            # calculate gates
+            gates = (torch.addmm(bias_batch, h_0, self.weight_hh) +
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply src/models/neural_networks/ealstm.py]: `i` is the input gate, and `g[t]` is the candidate cell input at timestep `t`. It goes "through" the input gate `i` (which you've correctly identified is simply element wise multiplication) before being added to the cell state.

File: tests/models/test_data.py (hunk 11 above)
Comment (gabrieltseng) [reply tests/models/test_data.py]: Yes

File: tests/models/neural_networks/test_ealstm.py (hunk 8 above)
Comment (gabrieltseng) [reply tests/models/neural_networks/test_ealstm.py]: Thank you! fixed

File: src/models/data.py [hunk 13]
```diff
@@ -230,15 +230,20 @@ def ds_folder_to_np(self,
         yearly_mean = x.mean(dim=['time', 'lat', 'lon'])
         yearly_mean_np = yearly_mean.to_array().values
 
+        yearly_std = x.std(dim=['time', 'lat', 'lon'])
+        yearly_std_np = yearly_std.to_array().values
+
+        yearly_agg = np.concatenate((yearly_mean_np, yearly_std_np))
```
Comment (tommylees112) [inline src/models/data.py]: what are the dimensions of this array then?

File: src/models/neural_networks/ealstm.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply src/models/neural_networks/ealstm.py]: fixed!

File: src/models/data.py (hunk 13 above)
Comment (gabrieltseng) [reply src/models/data.py]: `yearly_agg.shape == (num_variables * 2,)`

File: src/models/data.py (hunk 2 above)
Comment (tommylees112) [reply src/models/data.py]: gotcha!
